packages/react-components/react-utilities/src/hooks/useOnClickOutside.ts

```diff
@@ -42,18 +42,37 @@ export const useOnClickOutside = (options: UseOnClickOrScrollOutsideOptions) =>
   const timeoutId = React.useRef<number | undefined>(undefined);
   useIFrameFocus(options);
 
+  const mouseDownEventRef = React.useRef<MouseEvent | null>();
+
   const listener = useEventCallback((ev: MouseEvent | TouchEvent) => {
     const contains: UseOnClickOrScrollOutsideOptions['contains'] =
       containsProp || ((parent, child) => !!parent?.contains(child));
 
     const target = ev.composedPath()[0] as HTMLElement;
-    const isOutside = refs.every(ref => !contains(ref.current || null, target));
+
+    const isOutside = refs.every(ref => {
+      if (mouseDownEventRef.current) {
+        // Selecting text from inside to outside will rigger click event.
+        // In this case click event target is outside but mouse down event target is inside.
+        // And this click event should be considered as inside click.
+        return (
+          !contains(ref.current || null, target) &&
+          !contains(ref.current || null, mouseDownEventRef.current.target as HTMLElement)
+        );
+      }
+      return !contains(ref.current || null, target);
+    });
+    mouseDownEventRef.current = null;
 
     if (isOutside && !disabled) {
       callback(ev);
     }
   });
 
+  const handleMouseDown = useEventCallback((ev: MouseEvent) => {
```

Comment:
Do I understand correctly that if the mousedown event starts from inside, then it doesn't matter if the click it outside?

In this case you can simply check if the mousedown target is inside one of the refs with `.some` and then store a boolean in a ref so that the click listener can skip checking completely.

```ts
const isMouseDownInsideRect = React.useRef(false);

const onMouseDown = (e) => {
   isMouseDownInsideRef = refs.some(ref => contains(ref, e.target));
}

const onClick = () => {
   if (isMouseDownInsideRef.current) {
       // don't do anything
       return;
   }

   if(isOutside()) {
    // do smth
   }
}
```

Reply:
in the worst case you'd have to check inside/outside for all the elements twice... but the the code becomes much clearer

Reply:
The most common interaction is a simple click outside, and select text is less common. This way the code is cleaner but we'll almost always have to loop twice.

Reply:
Seeing how we use this utility, I don't think it's a big cost performance wise. We generally test for like 2 elements which is a trigger/container.

IMO the easier to understand code is much more useful for future maintainability than having 3 different contain statements in one check 🤷

I understand your argument so I won't block because of this though, it's your call

Reply:
Yeah I didn't realize refs is just an array of 2 elements. In this case I'm leaning to cleaner code, let me update it